tt_eager/tt_dnn/op_library/prod/kernels/dataflow/reader_prod_nc.cpp

```diff
@@ -0,0 +1,67 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdint>
+
+#include "dataflow_api.h"
+#include "tt_eager/tt_dnn/kernels/dataflow/moreh_common.hpp"
+
+inline uint32_t get_read_tile_id(uint32_t tile_id, uint32_t dim, uint32_t input_tile_offset, uint32_t HtWt) {
+    if(dim == 0){
+        return tile_id;
+    }else {
+        uint32_t a = 0 ;
+        while (tile_id >= HtWt){
+            tile_id-= HtWt;
+            a = a + 1;
+        }
+        uint32_t b = 0;
+        for (uint32_t i = 0; i < input_tile_offset; ++i) {
+            b = b + a;
+        }
+        return b + tile_id;
+    }
+}
```

Comment:
My suggestion to remove the div/mod was to integrate the accumulation into the main for loop by incrementing inside that loop, not writing your own div/mul routines.

Basic example:
```
constexpr uint32_t dim = compile_arg();
uint32_t batch_offset = get_arg_val();
uint32_t face_offset = get_arg_val();

for (...) {
    uint32_t read_tile_id;
    if constexpr (dim == 0) {
        read_tile_id = i;
    } else {
        read_tile_id = batch_offset + face_offset;
    }
    read_stuff;
    if constexpr (dim != 0) {
        face_offset++;
        if (face_offset == HtWt) {
            face_offset = 0;
            batch_offset += input_tile_offset;
        }
    }
}
```

Reply:
Sure